charmstore/fakeclient.go

```diff
@@ -0,0 +1,582 @@
+
+type miniHTTP interface {
+	Get(path string, value interface{}) error
+	Put(path string, value interface{}) error
+}
+
+type CharmStorer interface {
+	miniHTTP
+	UploadCharm(id *charm.URL, ch charm.Charm) (*charm.URL, error)
+	UploadCharmWithRevision(id *charm.URL, ch charm.Charm, promulgatedRevision int) error
+	UploadBundle(id *charm.URL, b charm.Bundle) (*charm.URL, error)
+	UploadBundleWithRevision(id *charm.URL, b charm.Bundle, promulgatedRevision int) error
+	UploadResource(id *charm.URL, name, path string, file io.ReadSeeker, size int64, progress csclient.Progress) (revision int, err error)
+	Publish(id *charm.URL, channels []params.Channel, resources map[string]int) error
+	AddDockerResource(id *charm.URL, resourceName string, imageName, digest string) (revision int, err error)
+	Latest(ids []*charm.URL, headers map[string][]string) ([]params.CharmRevision, error)
+	ListResources(id *charm.URL) ([]params.Resource, error)
+	GetResource(id *charm.URL, name string, revision int) (csclient.ResourceData, error)
+	ResourceMeta(id *charm.URL, name string, revision int) (params.Resource, error)
+	WithChannel(channel params.Channel) CharmStorer
+}
+
+// type ChannelAwareCharmStorer interface {
+// }
+
+// CharmFetcher provides access to charms and other objects
+// held in a charmstore (or locally), such as bundles and resources.
+//
+// c.f. github.com/juju/charmrepo.Interface and cmd/juju/application.DeployAPI
+type CharmFetcher interface {
+	// Get returns the charm referenced by charmURL.
+	Get(charmURL *charm.URL) (charm.Charm, error)
+
+	// GetBundle returns the bundle referenced by bundleURL.
+	GetBundle(bundleURL *charm.URL) (charm.Bundle, error)
+
+	// // GetResource returns the resource referenced by resourceURL.
+	// GetResource(resourceURL *charm.URL, name string, revision int) (csclient.ResourceData, error)
+
+	// ResolveWithChannel resolves an entity to its canonical form, providing the series that
+	// the entity supports as a slice of strings.
+	Resolve(entityURL *charm.URL) (canonicalURL *charm.URL, supportedSeries []string, err error)
+
+	// ResolveWithChannel resolves an entity to its canonical form, providing the series that
+	// the entity supports as a slice of strings. Unlike Resolve, ResolveWithChannel includes the
+	// entity's current channel in the result.
+	ResolveWithChannel(*charm.URL) (*charm.URL, params.Channel, []string, error)
+}
+
+// NoopProgress implements csclient.Progress
+type NoopProgress struct{}
+
+func (NoopProgress) Start(uploadId string, expires time.Time) {}
+func (NoopProgress) Transferred(total int64)                  {}
+func (NoopProgress) Error(err error)                          {}
+func (NoopProgress) Finalizing()                              {}
+
+var _ csclient.Progress = (*NoopProgress)(nil)
+
+func NewProgress() csclient.Progress {
+	return NoopProgress{}
+}
+
+// Datastore is a small in-memory key/value store that can be used
+// for faking HTTP calls.
+type Datastore struct {
```

Comment:
Why not just

```go
type Datastore map[string]interface{}
```
?

Reply:
Yeah, that's much cleaner. Using an inner field is a Rust-ism. Fixed.